src/main/java/org/broadinstitute/hellbender/tools/htsgetreader/HtsgetRequestBuilder.java

```diff
@@ -0,0 +1,217 @@
+    }
+
+    public void setFormat(final HtsgetFormat format) {
+        this.format = format;
+    }
+
+    public void setDataClass(final HtsgetClass dataClass) {
+        this.dataClass = dataClass;
+    }
+
+    public void setInterval(final SimpleInterval interval) {
+        this.interval = interval;
+    }
+
+    public void addField(final HtsgetRequestField field) {
+        this.fields.add(field);
+    }
+
+    public void addFields(final Collection<HtsgetRequestField> fields) {
+        this.fields.addAll(fields);
+    }
+
+    public void addTag(final String tag) {
+        this.tags.add(tag);
+    }
+
+    public void addTags(final Collection<String> tags) {
+        this.tags.addAll(tags);
+    }
+
+    public void addNotag(final String notag) {
+        this.notags.add(notag);
+    }
+
+    public void addNotags(final Collection<String> notags) {
+        this.notags.addAll(notags);
+    }
+
+    public HtsgetRequestBuilder withFormat(final HtsgetFormat format) {
+        this.format = format;
+        return this;
+    }
+
+    public HtsgetRequestBuilder withDataClass(final HtsgetClass dataClass) {
+        this.dataClass = dataClass;
+        return this;
+    }
+
+    public HtsgetRequestBuilder withInterval(final SimpleInterval interval) {
+        this.interval = interval;
+        return this;
+    }
+
+    public HtsgetRequestBuilder withField(final HtsgetRequestField field) {
+        this.fields.add(field);
+        return this;
+    }
+
+    public HtsgetRequestBuilder withFields(final Collection<HtsgetRequestField> fields) {
+        this.fields.addAll(fields);
+        return this;
+    }
+
+    public HtsgetRequestBuilder withTag(final String tag) {
+        this.tags.add(tag);
+        return this;
+    }
+
+    public HtsgetRequestBuilder withTags(final Collection<String> tags) {
+        this.tags.addAll(tags);
+        return this;
+    }
+
+    public HtsgetRequestBuilder withNotag(final String notag) {
+        this.notags.add(notag);
+        return this;
+    }
+
+    public HtsgetRequestBuilder withNotags(final Collection<String> notags) {
+        this.notags.addAll(notags);
+        return this;
+    }
+
+    /**
+     * Validates that the user query obeys htsget spec
+     */
+    private void validateRequest() {
+        if (this.dataClass != null && this.dataClass == HtsgetClass.header && (
+            this.interval != null ||
+            ! this.fields.isEmpty() ||
+            ! this.tags.isEmpty() ||
+            ! this.notags.isEmpty())) {
+                throw new UserException("Invalid request: no query parameters except `format` may be specified when class=header");
+        }
+
+        if (this.format != null) {
+            if ((this.id.endsWith(".bam") || this.id.endsWith(".cram") && (
+                this.format != HtsgetFormat.BAM && this.format != HtsgetFormat.CRAM))
+                ||
+                (this.id.endsWith(".vcf") || this.id.endsWith(".bcf")) && (
```

Comment:
do you need to include .vcf.gz as an option here?  There are a filename constants defined in  htsjdk.samtools.util.FileExtensions which are nice to reference instead of using raw strings. It make it easily searchable.

Reply:
Fixed.